### PR TITLE
[Merged by Bors] - Derive `Reflect` + `FromReflect` for input types

### LINKED ISSUE #6223: Add support for reflection to remaining input types
## What problem does this solve or what need does it fill?

This allows users to access the input types via reflection, which could be useful for scripting, debugging (via `bevy-inspector-egui` for example), etc.

## What solution would you like?

Add `#[derive(Reflect, FromReflect)]` to input types and register their types in the `InputPlugin`. Also add `#[reflect(Traits)]` for relevant traits to the types.

Common:
- `ButtonState`

Keyboard:
- `KeyCode` (`register_type` only, since it doesn't seem to be done in the `InputPlugin`)
- `KeyboardInput`

Mouse:
- `MouseButton`
- `MouseButtonInput`
- `MouseScrollUnit`
- `MouseMotion`
- `MouseWheel`

Gamepads:
- `GamepadButtonType`
- `GamepadAxisType`
- `Gamepad`
- `GamepadButton`
- `GamepadAxis`
- `GamepadEventType`
- `GamepadEvent`

Touch:
- `TouchPhase`
- `ForceTouch`
- `TouchInput`
- `Touch` (maybe?)
- `Touches` (maybe?)

For most (if not all) of these types, it's just a matter of adding `#[derive(Reflect, FromReflect)]` to the type and reflecting the relevant traits.

## What alternative(s) have you considered?

Users can recreate these types (with reflection support) in their own projects and create a system which maps the input types, but this step seems unnecessary. (This is different than mapping raw inputs to actions, it's mapping raw inputs to raw inputs.)

## Additional context

`KeyCode` already implements `Reflect` and `FromReflect`. The only change needed there is to add `.register_type::<KeyCode>()` in the `InputPlugin`.

### PR DESCRIPTION
**Pull request by TehPers**
# Objective

Adds support for reflecting many more of the input types. This allows those types to be used via scripting, `bevy-inspector-egui`, etc. These types are registered by the `InputPlugin` so that they're automatically available to anyone who wants to use them 

Closes #6223 

## Solution

Many types now have `#[derive(Reflect, FromReflect)]` added to them in `bevy_input`. Additionally, `#[reflect(traits...)]` has been added for applicable traits to the types.

This PR does not add reflection support for types which have private fields. Notably, `Touch` and `Touches` don't implement `Reflect`/`FromReflect`.

This adds the "glam" feature to the `bevy_reflect` dependency for package `bevy_input`. Since `bevy_input` transitively depends on `glam` already, all this brings in are the reflection `impl`s.

## Migration Guide

- `Input<T>` now implements `Reflect` via `#[reflect]` instead of `#[reflect_value]`. This means it now exposes its private fields via the `Reflect` trait rather than being treated as a value type. For code that relies on the `Input<T>` struct being treated as a value type by reflection, it is still possible to wrap the `Input<T>` type with a wrapper struct and apply `#[reflect_value]` to it.
  - As a reminder, private fields exposed via reflection are not subject to any stability guarantees.
---

## Changelog

Added
- Implemented `Reflect` + `FromReflect` for many input-related types. These types are automatically registered when adding the `InputPlugin`.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_input/src/gamepad.rs [hunk 1]
```diff
@@ -398,7 +407,8 @@ impl GamepadAxis {
 ///
 /// The [`GamepadSettings`] are used inside of the [`gamepad_event_system`], but are never written to
 /// inside of `bevy`. To modify these settings, mutate the corresponding resource.
-#[derive(Resource, Default, Debug)]
+#[derive(Resource, Default, Debug, Reflect, FromReflect)]
+#[reflect(Debug)]
```
Comment (MrGVSV) [inline crates/bevy_input/src/gamepad.rs:411]: You could also register the `ReflectDefault` type data.

File: crates/bevy_input/src/gamepad.rs [hunk 2]
```diff
@@ -484,7 +494,8 @@ impl GamepadSettings {
 /// ## Updating
 ///
 /// The current value of a button is received through the [`GamepadEvent`]s or [`GamepadEventRaw`]s.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Reflect, FromReflect)]
+#[reflect(Debug)]
```
Comment (MrGVSV) [inline crates/bevy_input/src/gamepad.rs:498]: Same here, you could register `ReflectDefault`.

File: crates/bevy_input/src/gamepad.rs [hunk 3]
```diff
@@ -534,7 +545,8 @@ impl ButtonSettings {
 /// ## Updating
 ///
 /// The current value of an axis is received through the [`GamepadEvent`]s or [`GamepadEventRaw`]s.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Reflect, FromReflect)]
+#[reflect(Debug)]
```
Comment (MrGVSV) [inline crates/bevy_input/src/gamepad.rs:549]: Same here, you could register `ReflectDefault`.

File: crates/bevy_input/src/gamepad.rs [hunk 4]
```diff
@@ -608,7 +620,8 @@ impl AxisSettings {
 /// ## Updating
 ///
 /// The current value of a button is received through the [`GamepadEvent`]s or [`GamepadEventRaw`]s.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Reflect, FromReflect)]
+#[reflect(Debug)]
```
Comment (MrGVSV) [inline crates/bevy_input/src/gamepad.rs:624]: Same here, you could register `ReflectDefault`.

File: crates/bevy_input/src/lib.rs [hunk 5]
```diff
@@ -84,12 +88,39 @@ impl Plugin for InputPlugin {
...
+            .register_type::<ScanCode>();
+        app.register_type::<MouseButtonInput>()
+            .register_type::<MouseButton>()
+            .register_type::<MouseMotion>()
+            .register_type::<MouseScrollUnit>()
+            .register_type::<MouseWheel>();
+        app.register_type::<TouchInput>()
+            .register_type::<ForceTouch>()
+            .register_type::<TouchPhase>();
+        app.register_type::<Gamepad>()
```
Comment (MrGVSV) [inline crates/bevy_input/src/lib.rs:105]: Super Minor Nit: I see why you did this. The grouping might be a bit clearer either with spaces between each group and/or a comment above each group.

For example:
```suggestion
        // Button types
        app.register_type::<ButtonState>();
        // Keyboard types
        app.register_type::<KeyboardInput>()
            .register_type::<KeyCode>()
            .register_type::<ScanCode>();
        // Mouse types
        app.register_type::<MouseButtonInput>()
            .register_type::<MouseButton>()
            .register_type::<MouseMotion>()
            .register_type::<MouseScrollUnit>()
            .register_type::<MouseWheel>();
        // Touch types
        app.register_type::<TouchInput>()
            .register_type::<ForceTouch>()
            .register_type::<TouchPhase>();
        // Gamepad types
        app.register_type::<Gamepad>()
```

File: crates/bevy_input/src/gamepad.rs (hunk 1 above)
Comment (TehPers) [reply crates/bevy_input/src/gamepad.rs:411]: I'll also add `ReflectSerialize` and `ReflectDeserialize` behind a `cfg_attr` where applicable. Thanks!

**Comment by TehPers** [comment]
Note: do not merge this yet. Currently, the `#[reflect(...)]` lists conflict with each other when the "serialize" feature is enabled. I left a comment about it in the Discord server, but basically only the last reflect list is used and the previous ones are discarded. We can mark this as a draft for the time being if anyone wants.

File: crates/bevy_input/src/input.rs [hunk 6]
```diff
@@ -35,7 +35,7 @@ use bevy_ecs::schedule::State;
 /// * Call the [`Input::release`] method for each release event.
 /// * Call the [`Input::clear`] method at each frame start, before processing events.
 #[derive(Debug, Clone, Resource, Reflect)]
-#[reflect_value]
+#[reflect_value(Default)]
```
Comment (MrGVSV) [inline crates/bevy_input/src/input.rs:38]: I might have already asked this, but does this need to be `reflect_value`? I feel like we'd potentially want access to the fields in the standard struct-like fashion.

Comment (MrGVSV) [reply crates/bevy_input/src/input.rs:38]: Yeah I think it's fine that the fields aren't public. Reflection provides a way to circumvent Rust's visibility system so it makes sense we can do it here.

File: crates/bevy_input/src/input.rs [hunk 7]
```diff
@@ -35,7 +35,7 @@ use bevy_ecs::schedule::State;
 /// * Call the [`Input::release`] method for each release event.
 /// * Call the [`Input::clear`] method at each frame start, before processing events.
 #[derive(Debug, Clone, Resource, Reflect)]
-#[reflect_value]
+#[reflect(Default)]
```
Comment (MrGVSV) [inline crates/bevy_input/src/input.rs:38]: Could you note this change inside a Migration Guide section in the PR description?

**Review by MrGVSV — APPROVED** [review]

**Comment by MrGVSV** [comment]
@TehPers Could you rebase this PR?

**Pull request closed without merging**